articles/ros2_dds_security.md

```diff
@@ -0,0 +1,220 @@
+It requires a public and private key per domain participant, as well as an x.509 certificate that binds the participant's public key to a specific name.
+Each x.509 certificate must be signed by (or have a signature chain to) a specific Certificate Authority (CA) that the plugin is configured to trust.
+
+
+## Access control
+
+The **Access control** plugin (see section 8.4 of the [DDS-Security spec][dds_security]) deals with defining and enforcing restrictions on the DDS-related capabilities of a given domain participant.
+For example, it allows a user to restrict a particular participant to a specific DDS domain, or only allow the participant to read from or write to specific DDS topics, etc.
+
+Again the SPI architecture allows for some flexibility in how the plugins accomplish this task, but ROS 2 uses the builtin access control plugin (called "DDS:Access:Permission", see section 9.4 of the [DDS-Security spec][dds_security]), which again uses PKI.
+It requires two files per domain participant:
+
+- **Governance** file: A signed XML document that specifies how the domain should be secured
+- **Permissions** file: A signed XML document that contains the permissions of the domain participant, bound to the name of the participant as defined by the authentication plugin (which is done via an x.509 cert, as we just discussed).
+
+Both of these files must be signed by a CA which the plugin is configured to trust.
+This may be the same CA that the **Authentication** plugin trusts, but that isn't required.
+
+
+## Cryptographic
+
+The **Cryptographic** plugin (see section 8.5 of the [DDS-Security spec][dds_security]) is where all the cryptography-related operations are handled: encryption, decryption, signing, hashing, etc.
+Both the **Authentication** and **Access control** plugins utilize the capabilities of the **Cryptographic** plugin in order to verify signatures, etc.
+This is also where the functionality to encrypt DDS topic communication resides.
+
+ROS 2 uses the builtin cryptographic plugin (called "DDS:Crypto:AES-GCM-GMAC", see section 9.5 of the [DDS-Security spec][dds_security]), which provides authenticated encryption using Advanced Encryption Standard (AES) in Galois Counter Mode (AES-GCM).
+
+
+# DDS-Security integration with ROS 2: SROS2
+
+Now that we have established some shared understanding of how security is supported in DDS, let's discuss how that support is exposed in ROS 2.
+By default, none of the security features of DDS are enabled in ROS 2.
+The set of features and tools in ROS 2 that are used to enable them are collectively named "Secure ROS 2" (SROS 2).
+
+
+## Features in the ROS client library (RCL)
+
+Most of the user-facing runtime support for SROS 2 is contained within the [ROS Client Library](https://github.com/ros2/rcl).
+Once its requirements are satisfied it takes care of configuring the middleware support for each supported DDS implementation.
```

Comment:
This is getting a bit pedantic, I'm sorry, but I'd say technically the actual configuring of the DDS-Security plugins is done in the rmw layer, e.g. https://github.com/ros2/rmw_fastrtps/blob/master/rmw_fastrtps_shared_cpp/src/rmw_node.cpp#L281, wouldn't you agree? RCL just passes the info along once it sorts out where everything is. As a result I think the current phrasing is more accurate than what you suggest.

Regarding the table, I think it's very helpful for understanding the current status, but I'm not sure it's particularly relevant to a document that doesn't actually discuss individual DDS vendors at all other than mentioning why all plugins aren't currently utilized by SROS 2. If you're prefer to see it here it would probably make the most sense above where we discuss which plugins are used. Thoughts?

Reply:
> but I'd say technically the actual configuring of the DDS-Security plugins is done in the rmw layer, e.g. https://github.com/ros2/rmw_fastrtps/blob/master/rmw_fastrtps_shared_cpp/src/rmw_node.cpp#L281, wouldn't you agree? RCL just passes the info along once it sorts out where everything is.

That's right

> As a result I think the current phrasing is more accurate than what you suggest.

Fair enough :+1:

> Regarding the table, I think it's very helpful for understanding the current status, but I'm not sure it's particularly relevant to a document that doesn't actually discuss individual DDS vendors at all other than mentioning why all plugins aren't currently utilized by SROS 2

Agreed, it's still unclear where such documentation would live that's why I suggested to add it to this "current state of DDS-Security integration document" but I agree that this level of implementation state detail does not belong in a design doc so let's leave it out :+1: